### DISCUSSION TITLE
Speaker labeling (recognition) in Rust

### DISCUSSION THREAD

**Discussion by thewh1teagle** [k2-fsa/sherpa-onnx#1077] [Q&A]
How can I use this library from Rust? I want to use the speaker recognition feature so I'll know who it talking and when.
As for the segmentaion, I'm thinking about using the word segments returning from whisper.cpp. is that good enough or it's bad to embed voice on each word?

**Comment by csukuangfj** [comment]
sorry, we don't support rust at present.

**Reply by thewh1teagle** [reply]
I'm on the way to create Rust bindings in [sherpa-rs](https://github.com/thewh1teagle/sherpa-rs)
Am I on the right track? does the c-api support all the APIs of sherpa?

**Reply by csukuangfj** [reply]
Cool!

Yes, C API is the recommended approach to use sherpa-onnx. If you find anything is missing in the C API, please let us know.

**Reply by thewh1teagle** [reply]
Thanks.
the embedding working on macOS :)
Now I'm trying to configure the build to work on Windows too. 
Building `sherpa-onnx` works on Windows, but when trying to build through the bindings it looks for `portaudio_static.lib` which should be in `build/lib` folder like the others lib files but it doesn't there either when build from bindings or just cloning the repo and buildings

Error from rust bindings build:
```console
           CMake Error at cmake_install.cmake:187 (file):
             file INSTALL cannot find
             "C:/Users/User/Documents/code/sherpa-rs/target/debug/build/sherpa-rs-sys-6a2d867f36688782/out/build/lib/Release/sherpa-onnx-portaudio_static.lib":
             No error.
```

Build command that works but doesn't produce that missing file:
```console
cmake -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=OFF -DSHERPA_ONNX_ENABLE_C_API=ON -DSHERPA_ONNX_ENABLE_WEBSOCKET=OFF -DSHERPA_ONNX_ENABLE_BINARY=OFF -DSHERPA_ONNX_ENABLE_TTS=OFF -DSHERPA_ONNX_ENABLE_PORTAUDIO=ON .
cmake --build build --config Release --parallel 8
```

Relevant file:
https://github.com/k2-fsa/sherpa-onnx/blob/master/cmake/portaudio.cmake

Not sure why it doesn't produce the static lib file
full log: https://gist.github.com/thewh1teagle/c983a32450c66ce7b8e1bac5593935d8

**Reply by csukuangfj** [reply]
Please set enable binary to ON.

I am refactoring how sherpa-onnx is built.

**Reply by thewh1teagle** [reply]
Also there's some issues with these two files in that folder (`main.go`, `run.sh`):
https://github.com/k2-fsa/sherpa-onnx/tree/master/scripts/go/_internal/vad-spoken-language-identification
Is it a sumlinks / submodules?
When trying to copy the whole folder in unix (macos) it throw error that these files not found

**Reply by csukuangfj** [reply]
please use go-api-examples

Everything inside scripts is for developers only.

**Reply by thewh1teagle** [reply]
I meaned that I copy the whole repo folder (to temp file when build in rust system). I have a workaround for that issue.
The [library](https://github.com/thewh1teagle/sherpa-rs) works and stable on macOS!
on Windows, I got code access violation error and warning  from sherpa

```console
PS C:\Users\User\Documents\code\sherpa-rs> cargo run --example speaker_id
   Compiling sherpa-rs-sys v0.1.0 (C:\Users\User\Documents\code\sherpa-rs\sys)
   Compiling sherpa-rs v0.1.0 (C:\Users\User\Documents\code\sherpa-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 37.23s
     Running `target\debug\examples\speaker_id.exe`
🎤 Loading model from C:\Users\User\Documents\code\sherpa-rs\nemo_en_speakerverification_speakernet.onnx
C:\Users\User\Documents\code\sherpa-rs\target\debug\build\sherpa-rs-sys-0bba50b59d1731c6\out\sherpa-onnx\sherpa-onnx\csrc\speaker-embedding-extractor.cc:Validate:29 Please provide --model
C:\Users\User\Documents\code\sherpa-rs\target\debug\build\sherpa-rs-sys-0bba50b59d1731c6\out\sherpa-onnx\sherpa-onnx\c-api\c-api.cc:SherpaOnnxCreateSpeakerEmbeddingExtractor:1123 Errors in config!
error: process didn't exit successfully: `target\debug\examples\speaker_id.exe` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)
```
The model file is there of course.
Looks like the expected model is `std::str` in cpp and the model string passed as empty string somehow?
This is how I pass it with `CString()` in [src/speaker_identify.rs#L29-L36](https://github.com/thewh1teagle/sherpa-rs/blob/main/src/speaker_identify.rs#L29-L36)

Also on macOS the provider ("cpu") passed as empty string too. but the model passed correctly

**Reply by csukuangfj** [reply]
Could you set
https://github.com/thewh1teagle/sherpa-rs/blob/main/src/speaker_identify.rs#L24
`debug` to true
and then post the logs again?

**Reply by thewh1teagle** [reply]
Sure

```console
 cargo run --example speaker_id      
   Compiling sherpa-rs-sys v0.1.0 (C:\Users\User\Documents\code\sherpa-rs\sys)
   Compiling sherpa-rs v0.1.0 (C:\Users\User\Documents\code\sherpa-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 26.80s
     Running `target\debug\examples\speaker_id.exe`
🎤 Loading model from C:\Users\User\Documents\code\sherpa-rs\nemo_en_speakerverification_speakernet.onnx
C:\Users\User\Documents\code\sherpa-rs\target\debug\build\sherpa-rs-sys-0bba50b59d1731c6\out\sherpa-onnx\sherpa-onnx\c-api\c-api.cc:SherpaOnnxCreateSpeakerEmbeddingExtractor:1119 SpeakerEmbeddingExtractorConfig(model="", num_threads=2, debug=True, provider="")

C:\Users\User\Documents\code\sherpa-rs\target\debug\build\sherpa-rs-sys-0bba50b59d1731c6\out\sherpa-onnx\sherpa-onnx\csrc\speaker-embedding-extractor.cc:Validate:29 Please provide --model
C:\Users\User\Documents\code\sherpa-rs\target\debug\build\sherpa-rs-sys-0bba50b59d1731c6\out\sherpa-onnx\sherpa-onnx\c-api\c-api.cc:SherpaOnnxCreateSpeakerEmbeddingExtractor:1123 Errors in config!
error: process didn't exit successfully: `target\debug\examples\speaker_id.exe` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)
```

Looks the same, note that I pass the debug option as i32 with value of `1`

**Reply by csukuangfj** [reply]
Could you change
https://github.com/k2-fsa/sherpa-onnx/blob/55decb7beea12a026b885bca1d2bc2dd306eed8f/sherpa-onnx/c-api/c-api.cc#L1118
to
```
if (true)
```
and retry and post the logs again?
(You need to re-built sherpa-onnx)

Looks like there are some issues with passing arguments from rust to c++.

**Reply by csukuangfj** [reply]
Oh, I see the logs now.
Please ignore my previous comments. (No need to change c-api.cc)
```
C:\Users\User\Documents\code\sherpa-rs\target\debug\build\sherpa-rs-sys-0bba50b59d1731c6\out\sherpa-onnx\sherpa-onnx\c-api\c-api.cc:
SherpaOnnxCreateSpeakerEmbeddingExtractor:1119 
SpeakerEmbeddingExtractorConfig(model="", num_threads=2, debug=True, provider="")
```

You can see that `model` is empty. Please re-check the arguments you passed from rust to C++.

**Reply by csukuangfj** [reply]
I am not familiar with rust.

In the following line
https://github.com/thewh1teagle/sherpa-rs/blob/main/src/speaker_identify.rs#L33
```
model: model_cstr.as_ptr(),
```

Will `model_cstr` be freed after the [function](https://github.com/thewh1teagle/sherpa-rs/blob/24ea7f5098f018db0ebbd9f3ea261f1f8c23ddee/src/speaker_identify.rs#L20) returns?

**Reply by thewh1teagle** [reply]
I think that it still manage it and freed it at the end of the function.
I changed it to `into_raw` instead of `as_ptr`, after reading that this way `Rust` "forget about it" and doesn't free it,
And then it worked!
How should I manage the memory in general with sherpa? does it correct to let rust forget about the memory in that case, and what in others?

**Reply by csukuangfj** [reply]
When you pass a pointer from rust to sherpa, you need to make sure the memory is not freed  before the function returns.

When you get a pointer from sherpa to rust, sherpa is responsible for managing the returned memory and you must not free it in rust.
You have to call sherpa's functions to free the returned pointer.

**Reply by csukuangfj** [reply]
>  does it correct to let rust forget about the memory in that case

I don't know what `forget` means in rust.

In c++, we use new and delete, or malloc and free to manage memory.

**Reply by thewh1teagle** [reply]
> When you pass a pointer from rust to sherpa, you need to make sure the memory is not freed before the function returns.

Simple! thanks.



> I don't know what `forget` means in rust.

I mean that it won't mess with the memory and free it after the function return.


> We support silero vad.

What should be the approach to implement diarization with sherpa? 
Is that correct:
1. get speech segments times using silero vad
2. Iterate segments and run one of `wespeaker` model for speaker identification (getting embeeding)
3. Check similarity to previous embedding and if bigger than N it's equal, otherwise push to "speakers" array (of embedding + ID)

I tried to run different models of speaker identifications but they returned different similarity values, beteewn 0 to 1, and it was very different between different audio files.
That's how I did it:
[examples/speaker_id.rs](https://github.com/thewh1teagle/sherpa-rs/blob/main/examples/speaker_id.rs)

**Reply by csukuangfj** [reply]
https://github.com/thewh1teagle/sherpa-rs/blob/fd4fd5c09197f58f93eb812eb89238b7387e48cf/src/speaker_identify.rs#L85

You have to free the ptr to avoid a memory leak. Please see the help doc about this function in c-api.h

**Reply by csukuangfj** [reply]
> but they returned different similarity values, beteewn 0 to 1, and it was very different between different audio files.

Could you explain this a bit? Sorry that I don't understand it.

**Reply by thewh1teagle** [reply]
> You have to free the ptr to avoid a memory leak. Please see the help doc about this function in c-api.h

The only c-api doc I can see is [onnx/c-api/index.html#generate-required-files](https://k2-fsa.github.io/sherpa/onnx/c-api/index.html#generate-required-files)



> Could you explain this a bit? Sorry that I don't understand it.

I plan to use sherpa for speaker diarization in the project [vibe](https://github.com/thewh1teagle/vibe) which uses whisper.cpp to create transcript, after it will done I want to segment the audio, and get speaker id for each sentence. I'm not entirely sure how I'm gonna do that.
I thought about iterate the segments, get embedding, check in existing embedding array if it's similar to one of them with cosine distance and if found then label the segment.

So I started test it with [speaker_identify.rs#L121](https://github.com/thewh1teagle/sherpa-rs/blob/main/src/speaker_identify.rs#L121) and the distances was very different in each segment sometimes even when it's the same speaker.
Anyway it can be other mistakes I did, but what do you think on the approach for diarization?

**Reply by csukuangfj** [reply]
please see
https://github.com/k2-fsa/sherpa-onnx/blob/55decb7beea12a026b885bca1d2bc2dd306eed8f/sherpa-onnx/c-api/c-api.h#L1002-L1004

It is clearly documented that you have to invoke `SherpaOnnxDestroySpeakerEmbeddingExtractor()`.

You can find how it is used in
1. go
https://github.com/k2-fsa/sherpa-onnx/blob/55decb7beea12a026b885bca1d2bc2dd306eed8f/scripts/go/sherpa_onnx.go#L986

2. C#
https://github.com/k2-fsa/sherpa-onnx/blob/55decb7beea12a026b885bca1d2bc2dd306eed8f/scripts/dotnet/SpeakerEmbeddingExtractor.cs#L34

3. JavaScript
https://github.com/k2-fsa/sherpa-onnx/blob/55decb7beea12a026b885bca1d2bc2dd306eed8f/scripts/node-addon-api/src/speaker-identification.cc#L70

There are plenty of examples of how to use our C API in different languages. I strongly recommend that you have a look at them.

**Reply by csukuangfj** [reply]
> but what do you think on the approach for diarization?

I think your approach is fine given that we currently only have vad + speaker embedding extractor models.

You can find an example of your approach in sherpa-onnx:
https://github.com/k2-fsa/sherpa-onnx/blob/master/python-api-examples/speaker-identification-with-vad-dynamic.py

Though it is in Python, you can change it to Rust if you want.

We have a demo video for it. Please see
https://www.bilibili.com/video/BV1Si421v7tk/

**Reply by thewh1teagle** [reply]
> It is clearly documented that you have to invoke `SherpaOnnxDestroySpeakerEmbeddingExtractor()`.

> There are plenty of examples of how to use our C API in different languages. I strongly recommend that you have a look at them.

Thanks! your suggestion was very good! now when continue developing it I look through the source code of the functions, and the bindings examples.



> I think your approach is fine given that we currently only have vad + speaker embedding extractor models.
> 
> You can find an example of your approach in sherpa-onnx:

I'm trying to use silero vad for segmenting audio (meaning getting start and stop positions of speech)
I created [examples/vad_segment.rs](https://github.com/thewh1teagle/sherpa-rs/blob/main/examples/vad_segment.rs) that uses [src/vad.rs#L62](https://github.com/thewh1teagle/sherpa-rs/blob/main/src/vad.rs#L62) but I'm not sure how to tune it so I'll get speech positions instead of silence.
Is there some example I follow? additionaly, then python examples have some defaults values which I don't find where they came from so I can set the same default in the Rust bindings. for example in python [config = sherpa_onnx.VadModelConfig()](https://github.com/k2-fsa/sherpa-onnx/blob/master/python-api-examples/vad-remove-non-speech-segments-from-file.py#L91) while in C bindings I need to fill [these values](https://github.com/thewh1teagle/sherpa-rs/blob/main/src/vad.rs#L16)

**Reply by csukuangfj** [reply]
> Is there some example I follow? additionaly, then python examples have some defaults values which I don't find where they came

I suggest that you use the same default value as our examples.

As said before, you can find usage of C APIs in C#, Go, JavaScript, etc. You can find default values there.

**Reply by thewh1teagle** [reply]
I tried to use the defaults from python, but looks like something is not correctly detected there.

The following example in python doesn't detect long clear and loud speech at the end of the audio with default config (Same in Rust)

```py
# wget https://github.com/snakers4/silero-vad/raw/master/files/silero_vad.onnx
# wget https://github.com/thewh1teagle/sherpa-rs/raw/main/samples/motivation.wav
# pip3 install soundfile numpy sherpa_onnx
# python3 main.py

from pathlib import Path
from typing import Tuple

import numpy as np
import sherpa_onnx
import soundfile as sf 


def load_audio(filename: str) -> Tuple[np.ndarray, int]:
    data, sample_rate = sf.read(
        filename,
        always_2d=True,
        dtype="float32",
    )
    data = data[:, 0]  # use only the first channel
    samples = np.ascontiguousarray(data)
    return samples, sample_rate

def main():
    samples, sample_rate = load_audio("motivation.wav")
    config = sherpa_onnx.VadModelConfig()
    config.silero_vad.model = "silero_vad.onnx"
    config.sample_rate = sample_rate

    window_size = config.silero_vad.window_size

    vad = sherpa_onnx.VoiceActivityDetector(config, buffer_size_in_seconds=30)
    while len(samples) > window_size:
        vad.accept_waveform(samples[:window_size])
        samples = samples[window_size:]
        if vad.is_speech_detected():
            while not vad.empty():
                start_sec = vad.front.start / sample_rate
                duration_sec = len(vad.front.samples) / sample_rate
                print(f"start={start_sec}s duration={duration_sec}s")
                vad.pop()

main()
```

**Reply by csukuangfj** [reply]
At the end of the audio, could you pad some zeros, e.g., 0.5 seconds of zeros.

**Reply by thewh1teagle** [reply]
I added padding of 1 second in the `load_audio` but still not detected at all speech between `13s-20s`
Maybe I didn't added them correctly, but also tried to add 1 seconds of silence with ffmpeg

Happens also with other files. after 10 seconds it doesn't detect speech

<details>
<summary>main.py</summary>

```py
# wget https://github.com/snakers4/silero-vad/raw/master/files/silero_vad.onnx
# wget https://github.com/thewh1teagle/sherpa-rs/raw/main/samples/motivation.wav
# pip3 install soundfile numpy sherpa_onnx
# python3 main.py

from pathlib import Path
from typing import Tuple

import numpy as np
import sherpa_onnx
import soundfile as sf 


def load_audio(filename: str) -> Tuple[np.ndarray, int]:
    data, sample_rate = sf.read(
        filename,
        always_2d=True,
        dtype="float32",
    )
    data = data[:, 0]  # use only the first channel
    samples = np.ascontiguousarray(data)
    
    # Add 1 seconds of padding to the end of the samples
    padding_samples = int(sample_rate * 1)
    samples = np.concatenate((samples, np.zeros(padding_samples, dtype=samples.dtype)))
    
    return samples, sample_rate

def main():
    samples, sample_rate = load_audio("motivation.wav")
    config = sherpa_onnx.VadModelConfig()
    config.silero_vad.model = "silero_vad.onnx"
    config.sample_rate = sample_rate

    window_size = config.silero_vad.window_size

    vad = sherpa_onnx.VoiceActivityDetector(config, buffer_size_in_seconds=3)
    while len(samples) > window_size:
        vad.accept_waveform(samples[:window_size])
        samples = samples[window_size:]
        if vad.is_speech_detected():
            while not vad.empty():
                start_sec = vad.front.start / sample_rate
                duration_sec = len(vad.front.samples) / sample_rate
                print(f"start={start_sec}s duration={duration_sec}s")
                vad.pop()

main()
```
</details>

**Reply by csukuangfj** [reply]
Could you upload the test wave for reproducing?
Please also describe the expected behavior.

**Reply by thewh1teagle** [reply]
The test file available here [samples/motivation.wav](https://github.com/thewh1teagle/sherpa-rs/raw/main/samples/motivation.wav) (`16khz`, `mono`)
When I run the attached script, it produce the following:

```console
python3 test.py 
start=0.926s duration=1.678s
start=3.774s duration=2.414s
start=7.518s duration=1.998s
```

The audio length is `20s` and there's loud and clear speech from `13s` to `20s` which is not detected.
expected behavior: it should find speech also in `13s` to `20s`

**Reply by thewh1teagle** [reply]
With original silero-vad it works correctly:

Result:
```console
python3 main.py
[{'start': 0.738, 'end': 2.43}, {'start': 3.81, 'end': 6.174}, {'start': 7.554, 'end': 9.15}, {'start': 12.77, 'end': 20.0}]
```

As you can see it detects speech from `12s` to `20s`. 


Code:
```python
# wget https://github.com/thewh1teagle/sherpa-rs/raw/main/samples/motivation.wav
# pip install torch torchaudio
# python3 main.py

import torch 
torch.set_num_threads(1)

model, utils = torch.hub.load(repo_or_dir='snakers4/silero-vad', model='silero_vad')
(get_speech_timestamps, _, read_audio, _, _) = utils

wav = read_audio('samples/motivation.wav')
sample_rate = 16000

def convert_samples_to_seconds(timestamps, sample_rate):
    return [{'start': ts['start'] / sample_rate, 'end': ts['end'] / sample_rate} for ts in timestamps]

speech_timestamps = get_speech_timestamps(wav, model)
readable_timestamps = convert_samples_to_seconds(speech_timestamps, sample_rate)
print(readable_timestamps)
```

Should I open new issue in sherpa-onnx?


Also, may be related, debug output:

```console
✗ cargo run --example diarize
   Compiling sherpa-rs v0.1.4-beta.1 (/Users/user/Documents/sherpa-rs)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 7.23s
     Running `target/debug/examples/diarize`
/Users/user/Documents/sherpa-rs/target/debug/build/sherpa-rs-sys-83bc881326eb63e6/out/sherpa-onnx/sherpa-onnx/c-api/c-api.cc:SherpaOnnxCreateVoiceActivityDetector:812 VadModelConfig(silero_vad=SilerVadModelConfig(model="silero_vad.onnx", threshold=0.5, min_silence_duration=0.4, min_speech_duration=0.4, window_size=512), sample_rate=16000, num_threads=1, provider="coreml", debug=True)
(speaker 0) start=0.936s duration=1.672s
(speaker 0) start=3.784s duration=2.408s
(speaker 0) start=7.528s duration=1.992s
/Users/user/Documents/sherpa-rs/target/debug/build/sherpa-rs-sys-83bc881326eb63e6/out/sherpa-onnx/sherpa-onnx/csrc/circular-buffer.cc:Push:96 Overflow! n: 512, size: 48000, n+size: 48512, capacity: 48000. Increase capacity to: 96000
/Users/user/Documents/sherpa-rs/target/debug/build/sherpa-rs-sys-83bc881326eb63e6/out/sherpa-onnx/sherpa-onnx/csrc/circular-buffer.cc:Push:96 Overflow! n: 512, size: 95616, n+size: 96128, capacity: 96000. Increase capacity to: 192000
```


Update:
I created new issue about it
